### ISSUE TITLE
Locals compaction on Power, AArch64 ?

### ISSUE THREAD

**Issue by 0xdaryl** [eclipse/openj9#5910]
I noticed that the JIT Power codegen doesn't support the compact locals optimization (where the JIT frame is compacted by sharing non-interfering autos).  Was that a conscious decision (e.g., it wouldn't be worth the effort because there are so many registers), or is it just an oversight?  @gita-omr @zl-wang 

Ditto for ARM and AArch64.

**Comment by zl-wang** [comment]
My impression is mainly the former (i.e., so many registers).  Plus, a comprehensive test was conducted re stack frame size around the time 64bit support was introduced ... even bigger stack frame in 64bit bore little portion of performance impact by 64bit: bigger object on cache capacity efficiency was largely the major performance issue.

**Comment by 0xdaryl** [comment]
If the last investigation was when 64-bit was introduced, that was 15 years ago.  I think the workloads of interest to OpenJ9 have changed, and profiles have gotten flatter and the stacks deeper.  Do you think it's worth repeating an investigation again?

**Comment by zl-wang** [comment]
It might be an interesting item for one of the students.  Let's gather some data wrt typically how many stack slots can be saved with compaction on etc, and its resulting performance impact.

**Comment by gita-omr** [comment]
Yes, I think it's worth trying. @0xdaryl you mean CompactLocals done by the optimizer?

**Comment by 0xdaryl** [comment]
Yes.  It's probably worth doing the investigation on either X or Z where the optimization is enabled.  Locals should be the same across architectures (i.e., it doesn't work on spill temps or other local artifacts created by backends).  On X86 there is some code in `OMR::Linkage::mapCompactedStack` guarded by DEBUG macros that can report some statistics on how the compaction went.  It's dusty code though, but probably can be resurrected pretty easily.

**Comment by fjeremic** [comment]
IMO if any work is done, this feature needs to sink down into OMR and be made cross platform. Nothing about local compaction is really codegen specific (may be some corner cases?). See eclipse/omr#5284 for details.

**Comment by 0xdaryl** [comment]
I agree, as much as possible.  There are some language-specific interference issues that may need to be abstracted away (e.g., can a reference and an integer share the same slot?  Some languages might say yes, others disallow it).

**Comment by knn-k** [comment]
@Akira1Saitoh has implemented and enabled locals compaction on AArch64 by https://github.com/eclipse/omr/pull/6387 and related PRs.
